language/move-lang/src/shared/mod.rs

```diff
@@ -0,0 +1,294 @@
+        write!(f, "{}", &self.value)
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for Spanned<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", &self.value)
+    }
+}
+
+/// Function used to have nearly tuple-like syntax for creating a Spanned
+pub const fn sp<T>(loc: Loc, value: T) -> Spanned<T> {
+    Spanned { loc, value }
+}
+
+/// Macro used to create a tuple-like pattern match for Spanned
+macro_rules! sp {
+    (_, $value:pat) => {
+        Spanned { value: $value, .. }
+    };
+    ($loc:pat, _) => {
+        Spanned { loc: $loc, .. }
+    };
+    ($loc:pat, $value:pat) => {
+        Spanned {
+            loc: $loc,
+            value: $value,
+        }
+    };
+}
+
+//**************************************************************************************************
+// Name
+//**************************************************************************************************
+
+// TODO maybe we should intern these strings somehow
+pub type Name = Spanned<String>;
+
+impl TName for Name {
+    type Key = String;
+    type Loc = Loc;
+
+    fn drop_loc(self) -> (Loc, String) {
+        (self.loc, self.value)
+    }
+
+    fn clone_drop_loc(&self) -> (Loc, String) {
+        (self.loc, self.value.clone())
+    }
+
+    fn add_loc(loc: Loc, key: String) -> Self {
+        sp(loc, key)
+    }
+}
+
+//**************************************************************************************************
+// Name
+//**************************************************************************************************
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct Counter(usize);
+
+impl Counter {
+    pub fn next() -> u64 {
+        static COUNTER_NEXT: AtomicUsize = AtomicUsize::new(0);
+
+        COUNTER_NEXT.fetch_add(1, AtomicOrdering::AcqRel) as u64
+    }
+}
+
+//**************************************************************************************************
+// Display
+//**************************************************************************************************
+
+pub fn display_comma<T: fmt::Display, I: Iterator<Item = T>>(
+    f: &mut fmt::Formatter,
+    mut items: I,
+) -> std::fmt::Result {
+    let first = items.next();
+    let x = match first {
+        None => return Ok(()),
+        Some(x) => x,
+    };
+    write!(f, "{}", x)?;
+    for item in items {
+        write!(f, ", {}", item)?;
+    }
+    Ok(())
+}
+
+pub fn format_comma<T: fmt::Display, I: Iterator<Item = T>>(items: I) -> String {
+    format_delim(items, ",")
+}
+
+pub fn format_delim<T: fmt::Display, I: Iterator<Item = T>>(mut items: I, delim: &str) -> String {
+    let mut res = String::new();
+    let first = items.next();
+    let x = match first {
+        None => return res,
+        Some(x) => x,
+    };
+    res.push_str(&format!("{}", x));
+    for item in items {
+        res.push_str(&format!("{} {}", delim, item));
+    }
+    res
+}
```

Comment:
Can't this be written as a `format!("{}", items.map(|t| format!("{}", t).collect().join(delim)))`